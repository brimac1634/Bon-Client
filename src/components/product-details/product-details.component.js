import React, { useState } from "react";
import { withRouter } from "react-router-dom";
import { connect } from "react-redux";
import {
  FacebookShareButton,
  TwitterShareButton,
  WhatsappShareButton,
  EmailShareButton
} from "react-share";

import { selectProduct } from "../../redux/shop/shop.selectors";
import { addItemStart } from "../../redux/cart/cart.actions";

import FormInput from "../form-input/form-input.component";
import CustomButton from "../custom-button/custom-button.component";
import HoverZoom from "../hover-zoom/hover-zoom.component";
import { ReactComponent as FacebookIcon } from "../../assets/facebook.svg";
import { ReactComponent as EmailIcon } from "../../assets/email.svg";
import { ReactComponent as TwitterIcon } from "../../assets/twitter.svg";
import { ReactComponent as WhatsappIcon } from "../../assets/whatsapp.svg";

import "./product-details.styles.scss";

const mapStateToProps = (state, ownProps) => ({
  product: selectProduct(ownProps.match.params.productID)(state),
  cartError: state.cart.error
});

const mapDispatchToProps = dispatch => ({
  addItem: item => dispatch(addItemStart(item))
});

export const ProductDetails = ({ product, addItem, cartItems }) => {
  const { name, images, price, description, features } = product;
  const [currentImage, setImage] = useState(images[0]);
  const [quantity, setQuantity] = useState(1);

  const quantityAvailable = product.quantity;

  const currentURL = window.location.href;

  return (
    <div className="product-details">
      <div className="panels">
        <div className="panel">
          <HoverZoom src={currentImage} alt="product image" />
          <div className="image-list">
            {images &&
              images.map((image, i) => (
                <div
                  key={i}
                  className="list-image"
                  style={{ backgroundImage: `url(${image})` }}
                  onClick={() => setImage(image)}
                />
              ))}
          </div>
        </div>
        <div className="panel">
          <h1>{name}</h1>
          <span className="price">{`HKD$${price}`}</span>
          {quantityAvailable >= 1 ? (
            <div className="order">
              <FormInput
                id="quantity"
                name="quantity"
                type="number"
                min={1}
                max={quantityAvailable}
                value={quantity}
                label="Quantity"
                handleChange={e => setQuantity(Number(e.target))}
              />
              <CustomButton
                onClick={() => addItem({ item: product, quantity })}
              >
                Add to Cart
              </CustomButton>
            </div>
          ) : (
            <span className="order not-avail">Out of Stock</span>
          )}

          <p className="description">{description}</p>
          {features &&
            features.map((feature, i) => (
              <div key={i} className="feature">
                <span>&#8226;</span>
                <span>{feature}</span>
              </div>
            ))}
          <div className="share">
            <span>Share:</span>
            <div className="share-buttons">
              <FacebookShareButton
                className="share-btn"
                url={currentURL}
                hashtag="#bonvivantcollection"
              >
                <div className="icon">
                  <FacebookIcon />
                </div>
              </FacebookShareButton>
              <TwitterShareButton
                className="share-btn"
                url={currentURL}
                hashtags={["bonvivantcollection"]}
              >
                <div className="icon">
                  <TwitterIcon />
                </div>
              </TwitterShareButton>
              <WhatsappShareButton
                className="share-btn"
                url={currentURL}
                title="Bon Vivant Collection"
              >
                <div className="icon">
                  <WhatsappIcon />
                </div>
              </WhatsappShareButton>
              <EmailShareButton
                className="share-btn"
                url={currentURL}
                subject={`Bon Vivant Collection - ${name}`}
              >
                <div className="icon">
                  <EmailIcon />
                </div>
              </EmailShareButton>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductDetails);
