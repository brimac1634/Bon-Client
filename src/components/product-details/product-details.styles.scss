.product-details {
	width: 100%;
	height: auto;
	padding: 40px 20px;
	display: flex;
	flex-direction: column;
	align-items: center;

	@media screen and (max-width: 450px) {
	    margin: 20px;
	}

	.panels {
		width: 100%;
		display: flex;
		justify-content: space-around;
		flex-wrap: wrap;

		.panel {
			width: 100%;
			max-width: 450px;
			display: flex;
			flex-direction: column;
			align-items: flex-start;
			margin-bottom: 40px;

			h1 {
				margin-top: 0;
			}

			span, p {
				font-size: 120%;
			}

			.order {
				margin: 40px 0 20px 0;

				&.not-avail {
					font-weight: 600;
					font-size: 150%;
				}
			}

			.image-list {
				width: 100%;
				display: grid;
				grid-template-columns: repeat(4, 1fr);
			    grid-gap: 10px;

				.list-image {
					width: 100%;
					border: solid 1px grey;
					background-size: cover;
					background-position: center;
					cursor: pointer;
					transition: all 200ms ease-in-out;

					&:after {
						content: "";
						display: block;
						padding-bottom: 100%;
					}

					&:hover {
				        opacity: 0.8;
				    }
				}
			}

			.price {
				color: grey;
				font-size: 150%;
			}

			.description {
				margin: 20px 0;
			}

			.feature {
				display: flex;

				:first-child {
					margin: 0 10px 5px 0;
				}
			}

			.share {
				margin-top: 40px;
				display: flex;
				flex-direction: column;
				align-items: flex-start;

				:first-child {
					margin-bottom: 10px;
				}
				
				.share-buttons {
					display: flex;
					cursor: pointer;

					&:hover .icon {
						fill: gray;
					}

					.icon {
						margin: 0 5px;
						display: flex;
						justify-content: center;
						align-items: center;

						&:hover {
							fill: black;
						}

						svg {
							width: 26px;
							height: 26px;
							transition: all 200ms ease-in-out;
							margin: 0;
						}
						
					}
				}
			}
		}
	}

	
}