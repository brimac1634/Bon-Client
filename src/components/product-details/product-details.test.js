import { shallow } from "enzyme";
import React from "react";
import { ProductDetails } from "./product-details.component";

describe("<ProductDetails />", () => {
  const mockProduct = {
    name: "test product",
    images: ["test_image"],
    price: 60,
    description: "a test product",
    features: ["something nice"],
    quantity: 2
  };

  const mockStore = {
    product: mockProduct,
    cartError: null
  };

  const wrapper = shallow(
    <ProductDetails store={mockStore} product={mockProduct} />
  );

  it("expect to render ProductDetails component", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  // it('expect quantity input to increment', () => {

  // 	wrapper.find('[id="quantity"]').simulate('change', {target: {value: 5}})
  // 	expect(wrapper.state()).toEqual({ quantity: 5 })
  // })
});
