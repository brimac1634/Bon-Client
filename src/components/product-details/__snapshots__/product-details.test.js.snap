// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetails /> expect to render ProductDetails component 1`] = `
"<div className=\\"product-details\\">
  <div className=\\"panels\\">
    <div className=\\"panel\\">
      <HoverZoom src=\\"test_image\\" alt=\\"product image\\" />
      <div className=\\"image-list\\">
        <div className=\\"list-image\\" style={{...}} onClick={[Function: onClick]} />
      </div>
    </div>
    <div className=\\"panel\\">
      <h1>
        test product
      </h1>
      <span className=\\"price\\">
        HKD$60
      </span>
      <div className=\\"order\\">
        <FormInput id=\\"quantity\\" name=\\"quantity\\" type=\\"number\\" min={1} max={2} value={1} label=\\"Quantity\\" handleChange={[Function: handleChange]} />
        <CustomButton onClick={[Function: onClick]}>
          Add to Cart
        </CustomButton>
      </div>
      <p className=\\"description\\">
        a test product
      </p>
      <div className=\\"feature\\">
        <span>
          â€¢
        </span>
        <span>
          something nice
        </span>
      </div>
      <div className=\\"share\\">
        <span>
          Share:
        </span>
        <div className=\\"share-buttons\\">
          <ForwardRef className=\\"share-btn\\" url=\\"http://localhost/\\" hashtag=\\"#bonvivantcollection\\" windowWidth={550} windowHeight={400}>
            <div className=\\"icon\\">
              <ForwardRef(SvgFacebook) />
            </div>
          </ForwardRef>
          <ForwardRef className=\\"share-btn\\" url=\\"http://localhost/\\" hashtags={{...}} windowWidth={550} windowHeight={400}>
            <div className=\\"icon\\">
              <ForwardRef(SvgTwitter) />
            </div>
          </ForwardRef>
          <ForwardRef className=\\"share-btn\\" url=\\"http://localhost/\\" title=\\"Bon Vivant Collection\\" separator=\\" \\" windowWidth={550} windowHeight={400}>
            <div className=\\"icon\\">
              <ForwardRef(SvgWhatsapp) />
            </div>
          </ForwardRef>
          <ForwardRef className=\\"share-btn\\" url=\\"http://localhost/\\" subject=\\"Bon Vivant Collection - test product\\" separator=\\" \\" openWindow={false} onClick={[Function: onClick]}>
            <div className=\\"icon\\">
              <ForwardRef(SvgEmail) />
            </div>
          </ForwardRef>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
